modules/tinymce/src/core/main/ts/fmt/FormatChanged.ts

```diff
@@ -5,101 +5,167 @@
+  Arr.findUntil(
+    parents,
+    (node) => { // is this the node we want?
+      const matchingFormat = editor.formatter.matchNode(node, format, vars ?? {}, similar);
+      return !Type.isUndefined(matchingFormat);
+    },
+    (node) => { // Can we guarantee that no parent of this will be the node we want?
+      if (MatchFormat.matchesUnInheritedFormatSelector(editor, node, format)) {
+        return true;
+      } else {
+        if (!similar) {
+          // If we want to find an exact match, then finding a similar match halfway up the parents tree is bad
+          return Type.isNonNullable(editor.formatter.matchNode(node, format, vars, true));
+        } else {
+          return false;
+        }
+      }
+    }
+  );
+
+const getParents = (editor: Editor, elm?: Element): Element[] => {
+  const element = elm ?? editor.selection.getNode();
+  return Arr.filter(FormatUtils.getParents(editor.dom, element), (node) =>
+    node.nodeType === 1 && !node.getAttribute('data-mce-bogus')
+  );
+};
+
+const hasChanged = (state: Cell<boolean>, newState: boolean): boolean => {
+  if (state.get() !== newState) {
+    state.set(newState);
+    return true;
+  } else {
+    return false;
+  }
+};
 
+const updateAndFireChangeCallbacks = (
+  editor: Editor,
+  elm: Element,
+  registeredCallbacks: RegisteredFormats
+) => {
   // Ignore bogus nodes like the <a> tag created by moveStart()
-  const parents = Arr.filter(FormatUtils.getParents(editor.dom, elm), (node) => node.nodeType === 1 && !node.getAttribute('data-mce-bogus'));
-
-  // Check for new formats
-  Obj.each(formatChangeData, (data: FormatData, format: string) => {
-    Tools.each(parents, (node: Node) => {
-      if (editor.formatter.matchNode(node, format, {}, data.similar)) {
-        if (formatsList.indexOf(format) === -1) {
-          // Execute callbacks
-          Arr.each(data.callbacks, (callback: FormatChangeCallback) => {
-            callback(true, { node, format, parents });
-          });
-
-          newFormats[format] = data.callbacks;
-        }
+  const parents = getParents(editor, elm);
+
+  Obj.each(registeredCallbacks, (data: CallbackGroup, format: string) => {
+    if (data.similarCallbacks.length > 0) {
+      const match = matchingNode(editor, parents, format, true);
+      if (hasChanged(data.similarState, match.isSome())) {
+        const node = match.getOr(elm);
+        Arr.each(data.similarCallbacks, (callback) => callback(match.isSome(), { node, format, parents }));
+      }
+    }
 
-        matchedFormats[format] = data.callbacks;
-        return false;
+    if (data.similarFalseCallbacks.length > 0) {
+      const match = matchingNode(editor, parents, format, false);
+      if (hasChanged(data.similarFalseState, match.isSome())) {
+        const node = match.getOr(elm);
+        Arr.each(data.similarFalseCallbacks, (callback) => callback(match.isSome(), { node, format, parents }));
       }
+    }
```

Comment:
Maybe a function here as we have similar code?!

Reply:
Have refactored this, but I'm not sure if the way I did it was good or not. Can someone take another look?